### PR TITLE
[Documentation] Outline Elastic IP Manager Design

### PR DESCRIPTION
**Pull request by xieus**
This PR offers a design outline of Elastic IP Manager to kick off discussion.

### CODE REVIEW

File: docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+* Associate or disassociate an EIP to a running instance, which by default attach to its primary interface.
+* Associate or disassociate an EIP to a network interface.
+
+=== Service Basic Requirements
+
+[arabic]
+1. Provide CURD REST APIs for customers or provider admins to management EIPs. The APIs are compatible to OpenStack Neutron v2.0 APIs <<neutron_api>>.
+1. Provide CURD REST APIs for cloud provider admins to register or unregister a list of public IPv4 or IPv6 ranges in EIP manager.
+Note that those IP ranges should be owned by cloud provider.
+1. Allocate _standalone_ and _unique_ EIP address. Guarantee uniqueness of EIP address in a regional scope.
```
Comment (er1cthe0ne) [inline docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc:28]: should the uniqueness be in the world level?

Comment (xieus) [reply docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc:28]: An EIP is a resource unique globally but the EIP service in one region has no visibility on other EIP services. Therefore from EIP service perspective, it only guarantee region-wide uniqueness while the global uniqueness is ensured by additional mechanisms.

**Comment by xieus** [comment]
@Eric-Yuan see if you have additional comments.

File: docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+1. Release public IP address per customer request or other microservices, and make it ready for reuse for other customers.
+1. Release (port, EIP) association when a port is to be deleted.
+1. Maintain EIP states and a mapping from an EIP to its owning interface/port.
+1. Association of an EIP to a running instance or network interface (port) at POST call (when fixed_ip_address is not NULL), or at PUT call.
+1. Disassociation of EIP with a network interface (primary or secondary).
+1. Working together with other services including port manager and DNS manager for allocation of port and external DNS, respectively.
```
[the remark this answers is not in the document]
Comment (Eric-Yuan) [reply docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc:34]: I saw port manger has already support dns_name and dns_assignment attribute, so mayb elastic IP manager should support that too.

File: docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+The Elastic IP (EIP) manager is responsible of allocating and releasing EIP as well as associating the allocated EIP to a running instance or network interface (port).
+This is the acting microservice that provides elastic/floating IP APIs to external customers (through API gateway) and driving the elastic IP CURD workflow.
+A network interface with an enabled EIP could establish inbound connections from and outbound connections to external networks like Internet.
+
+=== User Cases
+A customer of Alcor could leverage Elastic IP manager in the following cases:
+
+* Allocate or release a _standalone_ EIP within the user account.
+* Associate or disassociate an EIP to a running instance, which by default attach to its primary interface.
```
[the remark this answers is not in the document]
Comment (xieus) [reply docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc:19]: This is more like customer scenario. From EIP manager perspective, the EIP is associated to the primary interface attached to the running instance. EIP manager as of now is unaware of primary or secondary interface therefore the API exposed by EIP manager is generic to either primary interface or secondary interface.

File: docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (xieus) [reply docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc:34]: Correct. Eventually EIP manager could talk to external and/or internal DNS manager to allocate DNS name etc. for an EIP. We could split the DNS management into two microservices, one for external DNS and the other for internal DNS, or consolidate them into one bigger microservice. Either approach has its pros and cons. 

We leave some design space in EIP manager for its interaction with DNS management component, and revisit it shortly after DNS design is finalized.

File: docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (Eric-Yuan) [reply docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc:19]: Get it

File: docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Eric-Yuan) [reply docs/modules/ROOT/pages/mgmt_services/elastic_ip_manager.adoc:34]: OK

**Pull request merged**